[scroll-anchoring] Implement suppression triggers

nmoucht opened the pull request:
#### b02b493840f0ce869e34fcd65a027245d392bc91
<pre>
[scroll-anchoring] Implement suppression triggers
<a href="https://bugs.webkit.org/show_bug.cgi?id=261719">https://bugs.webkit.org/show_bug.cgi?id=261719</a>
<a href="https://rdar.apple.com/115704143">rdar://115704143</a>

Reviewed by Simon Fraser.

Implement suppression of scroll anchoring adjustments according to the spec
(<a href="https://www.w3.org/TR/css-scroll-anchoring-1/#suppression-triggers).">https://www.w3.org/TR/css-scroll-anchoring-1/#suppression-triggers).</a> This involves suppressing scroll
 anchoring adjustments  when certain “suppression trigger” operations occur when an anchor element has
 been chosen but before a scroll anchoring adjustment has occurred, that would cause the next scroll
anchoring adjustment to be incorrect. These suppression triggers are certain style changes on the anchor
element or any element in the anchor element ancestor chain, up to and including the scrolling element
that owns the scroll anchroing controller, as well as changing to or from being absolutely posititioned,
for any element under the owning scrolling element.

* LayoutTests/imported/w3c/web-platform-tests/css/css-scroll-anchoring/ancestor-change-heuristic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scroll-anchoring/heuristic-with-offset-update-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scroll-anchoring/opt-out-dynamic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scroll-anchoring/opt-out-dynamic-scroller-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scroll-anchoring/position-change-heuristic-expected.txt:
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::scheduleResizeEventIfNeeded):
* Source/WebCore/page/scrolling/ScrollAnchoringController.cpp:
(WebCore::elementIsScrollableArea):
(WebCore::setInScrollAnchoringAncestorChain):
(WebCore::ScrollAnchoringController::invalidateAnchorElement):
(WebCore::ScrollAnchoringController::notifyChildHadSuppressingStyleChange):
(WebCore::scrollAnchoringControllerForElement):
(WebCore::ScrollAnchoringController::notifyParentScrollAnchoringControllerHadSuppressingStyleChange):
(WebCore::ScrollAnchoringController::didFindPriorityCandidate):
(WebCore::ScrollAnchoringController::chooseAnchorElement):
(WebCore::ScrollAnchoringController::adjustScrollPositionForAnchoring):
* Source/WebCore/page/scrolling/ScrollAnchoringController.h:
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::styleDidChange):
* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::isInScrollAnchoringAncestorChain const):
(WebCore::RenderObject::setIsInScrollAnchoringAncestorChain):
(WebCore::RenderObject::RenderObjectBitfields::RenderObjectBitfields):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::scrollAnchoringSuppressionStyleDidChange const):
(WebCore::RenderStyle::absolutePositionStyleDidChange const):
* Source/WebCore/rendering/style/RenderStyle.h:

Canonical link: <a href="https://commits.webkit.org/270455@main">https://commits.webkit.org/270455@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/71c6a63db64514e25666fbd0c93a7f85d2e7d2f0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/25493 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4085 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/26768 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/27596 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23363 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/25772 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/5822 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/1520 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/23527 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/25738 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/3016 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/21976 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/28177 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/2668 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/22925 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/29029 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/23244 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/23288 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/26853 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/2664 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/917 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4045 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3118 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3261 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3003 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/page/scrolling/ScrollAnchoringController.cpp:
```diff
@@ -85,7 +112,40 @@ FloatPoint ScrollAnchoringController::computeOffsetFromOwningScroller(RenderObje
+
+static ScrollAnchoringController* scrollAnchoringControllerForElement(Element* element)
+{
+    if (element->renderBox() && element->renderBox()->layer() && element->renderBox()->layer()->scrollableArea() && element->renderBox()->layer()) {
```
This should use either `enclosingLayer` or code shared with `EventHandler::enclosingScrollableArea()`. Does the spec require that this ancestor finding is a DOM ancestor walk or is it OK to walk the ancestor chain via renderers?

nmoucht replied:
Removed this in favor of finding the anchoring element via RenderObject. I think it is reasonable in this case to use a the render chain to find the nearest scrollable area.

smfr commented on Source/WebCore/page/scrolling/ScrollAnchoringController.cpp:
```diff
@@ -43,19 +43,46 @@ ScrollAnchoringController::ScrollAnchoringController(ScrollableArea& owningScrol
 
+ScrollAnchoringController::~ScrollAnchoringController()
+{
+    invalidateAnchorElement();
```
Be careful adding code to a destructor like this which calls out via `setInScrollAnchoringAncestorChain()` to poke things on the render tree. What does it do when the entire render tree is being torn down?

nmoucht replied:
Removed the setInScrollAnchoringAncestorChain call but I think we should still invalidate the current anchor so that the set of scrollable areas needing update on the FrameView is correct.

smfr commented on Source/WebCore/page/scrolling/ScrollAnchoringController.cpp:
```diff
@@ -43,19 +43,46 @@ ScrollAnchoringController::ScrollAnchoringController(ScrollableArea& owningScrol
 
+static bool elementIsScrollableArea(Element* element, ScrollableArea& scrollableArea)
+{
+    return element->renderBox() && element->renderBox()->layer() && element->renderBox()->layer()->scrollableArea() && (element->renderBox()->layer()->scrollableArea()->scrollingNodeID() == scrollableArea.scrollingNodeID());
```
Not sure why you compare `scrollingNodeID`. Why not just compare scrollableArea by address? Also, maybe call `canBeScrolledAndHasScrollableArea()`?

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/rendering/RenderObject.h:
```diff
@@ -402,6 +402,9 @@ class RenderObject : public CachedImageClient, public CanMakeCheckedPtr {
 
     bool everHadLayout() const { return m_bitfields.everHadLayout(); }
 
+    bool isInScrollAnchoringAncestorChain() const { return m_bitfields.isInScrollAnchoringAncestorChain(); }
```
I think we should try an implementation that doesn't store this hard-to-maintain bit on RenderObject. That means we need to write an efficient `isInScrollAnchoringAncestorChain()` since this will be hot in the style changed code path.

I think we can. From a RenderObject you can get its enclosingLayer. On a RenderLayer you can look at its `ScrollingScope` to see if it's in an async-scrollable overflow which is not the root (note; if we have to scroll-anchor inside non-async scrollables, you'll need to write more code). You can then fetch the scroll anchoring node of the relevant root or overflow scroll, and use your `setInScrollAnchoringAncestorChain()` logic.

nmoucht replied:
Changed to getting the scroll anchoring controller via render ancestor walk, then using the scroll anchor controller to check if the render object is in a scroll anchoring ancestor change via `isInScrollAnchoringAncestorChain`.

smfr commented on Source/WebCore/rendering/style/RenderStyle.cpp:
```diff
@@ -1376,6 +1377,49 @@ bool RenderStyle::changeRequiresRecompositeLayer(const RenderStyle& other, Optio
     return false;
 }
 
+bool RenderStyle::scrollAnchoringSuppressionStyleDidChange(const RenderStyle* other) const
```
Would be good to link to the spec text for this block.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/rendering/style/RenderStyle.cpp:
```diff
@@ -1376,6 +1377,49 @@ bool RenderStyle::changeRequiresRecompositeLayer(const RenderStyle& other, Optio
+        return true;
+    if (m_nonInheritedFlags.position != other->m_nonInheritedFlags.position)
+        return true;
+    if (m_nonInheritedData->surroundData->margin != other->m_nonInheritedData->surroundData->margin)
```
Better to use use `margin()` directly unless you're going to add optimizations about checking the m_nonInheritedData or surroundData pointers.

nmoucht replied:
Changed to check the surroundData pointers.

smfr commented on Source/WebCore/rendering/style/RenderStyle.cpp:
```diff
@@ -1376,6 +1377,49 @@ bool RenderStyle::changeRequiresRecompositeLayer(const RenderStyle& other, Optio
+            return true;
+    }
+    if (overflowAnchor() != other->overflowAnchor() && overflowAnchor() == OverflowAnchor::None)
+        return true;
```
Be consistent about blank lines after a return (I prefer them).

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/rendering/style/RenderStyle.cpp:
```diff
@@ -1376,6 +1377,49 @@ bool RenderStyle::changeRequiresRecompositeLayer(const RenderStyle& other, Optio
+bool RenderStyle::absolutePositionStyleDidChange(const RenderStyle* other) const
+{
+    if (other && m_nonInheritedData.ptr() != other->m_nonInheritedData.ptr()) {
+        if (hasOutOfFlowPosition() != other->hasOutOfFlowPosition())
```
This is affected by `position:fixed`  but the spec only refers to `position:absolute` changes. Again, a spec link would be useful.

nmoucht replied:
Added spec link but I think its referring to both position:fixed and position:absolute elements when it mentions absolutely positioned elements.

smfr commented on Source/WebCore/rendering/style/RenderStyle.h:
```diff
@@ -2135,6 +2135,8 @@ class RenderStyle {
     inline BlockStepInsert blockStepInsert() const;
     inline void setBlockStepInsert(BlockStepInsert);
+    bool scrollAnchoringSuppressionStyleDidChange(const RenderStyle*) const;
+    bool absolutePositionStyleDidChange(const RenderStyle*) const;
```
Let's say "outOfFlowPositionStyleDidChange" otherwise I will always think this excludes `fixed`.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/ScrollAnchoringController.cpp:
```diff
@@ -43,13 +43,23 @@ ScrollAnchoringController::ScrollAnchoringController(ScrollableArea& owningScrol
     return downcast<LocalFrameView>(downcast<ScrollView>(m_owningScrollableArea));
 }
 
+static bool elementIsScrollableArea(Element* element, ScrollableArea& scrollableArea)
```
This should take an `Element&` since it's assumed never to be null. Both args can be `const`.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/ScrollAnchoringController.cpp:
```diff
@@ -43,13 +43,23 @@ ScrollAnchoringController::ScrollAnchoringController(ScrollableArea& owningScrol
 
+static bool elementIsScrollableArea(Element* element, ScrollableArea& scrollableArea)
+{
+    return element->renderBox() && element->renderBox()->layer() && element->renderBox()->layer()->scrollableArea() && (element->renderBox()->layer()->scrollableArea() == &scrollableArea);
```
You don't need to check for a non-null `element->renderBox()->layer()->scrollableArea()` because the comparison with the never-null `scrollableArea` will take care of that.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/ScrollAnchoringController.cpp:
```diff
@@ -85,6 +95,30 @@ FloatPoint ScrollAnchoringController::computeOffsetFromOwningScroller(RenderObje
+    m_shouldSupressScrollPositionUpdate = true;
+}
+
+bool ScrollAnchoringController::isInScrollAnchoringAncestorChain(const RenderObject* object)
```
Looks like the argument is never null so should be a reference.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/ScrollAnchoringController.cpp:
```diff
@@ -249,15 +283,20 @@ void ScrollAnchoringController::updateAnchorElement()
+    auto supressed = std::exchange(m_shouldSupressScrollPositionUpdate, false);
     if (!m_anchorElement || !queued)
         return;
     auto renderBox = m_anchorElement->renderer();
```
`renderer()` returns a `RenderObject`, not a `RenderBox`, so calling this `renderBox` is misleading.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/rendering/RenderObject.cpp:
```diff
@@ -2349,6 +2350,22 @@ Vector<FloatRect> RenderObject::clientBorderAndTextRects(const SimpleRange& rang
+        if (auto* scrollableArea = downcast<RenderLayerModelObject>(*this).layer()->scrollableArea())
+            return scrollableArea->scrollAnchoringController();
+    }
+    for (auto* renderer = this; renderer; renderer = renderer->parent()) {
```
This loop would be faster if written in terms of `RenderLayer`. So call `enclosingLayer()` and then walk up the layer parent chain.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/rendering/RenderObject.h:
```diff
@@ -402,6 +402,8 @@ class RenderObject : public CachedImageClient, public CanMakeCheckedPtr {
 
     bool everHadLayout() const { return m_bitfields.everHadLayout(); }
 
+    ScrollAnchoringController* scrollAnchoringController() const;
```
Having this here makes it sound like each RenderObject has one, but this is really a "find one by traversing ancestors", so the name should reflect that.

It would even be better as a static function: `static ScrollAnchoringController* findScrollAnchoringControllerForRenderer(const RenderObject&)`.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/rendering/style/RenderStyle.cpp:
```diff
@@ -1376,6 +1377,60 @@ bool RenderStyle::changeRequiresRecompositeLayer(const RenderStyle& other, Optio
+    if (position() != other->position())
+        return true;
+
+    if (m_nonInheritedData->surroundData != other->m_nonInheritedData->surroundData) {
```
Should this check `m_nonInheritedData->surroundData.ptr()`?

nmoucht replied:
Fixed.

nmoucht committed:
[scroll-anchoring] Implement suppression triggers
https://bugs.webkit.org/show_bug.cgi?id=261719
rdar://115704143

Reviewed by Simon Fraser.

Implement suppression of scroll anchoring adjustments according to the spec
(https://www.w3.org/TR/css-scroll-anchoring-1/#suppression-triggers). This involves suppressing scroll
 anchoring adjustments  when certain “suppression trigger” operations occur when an anchor element has
 been chosen but before a scroll anchoring adjustment has occurred, that would cause the next scroll
anchoring adjustment to be incorrect. These suppression triggers are certain style changes on the anchor
element or any element in the anchor element ancestor chain, up to and including the scrolling element
that owns the scroll anchroing controller, as well as changing to or from being absolutely posititioned,
for any element under the owning scrolling element.

* LayoutTests/imported/w3c/web-platform-tests/css/css-scroll-anchoring/ancestor-change-heuristic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scroll-anchoring/heuristic-with-offset-update-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scroll-anchoring/opt-out-dynamic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scroll-anchoring/opt-out-dynamic-scroller-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scroll-anchoring/position-change-heuristic-expected.txt:
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::scheduleResizeEventIfNeeded):
* Source/WebCore/page/scrolling/ScrollAnchoringController.cpp:
(WebCore::elementIsScrollableArea):
(WebCore::setInScrollAnchoringAncestorChain):
(WebCore::ScrollAnchoringController::invalidateAnchorElement):
(WebCore::ScrollAnchoringController::notifyChildHadSuppressingStyleChange):
(WebCore::scrollAnchoringControllerForElement):
(WebCore::ScrollAnchoringController::notifyParentScrollAnchoringControllerHadSuppressingStyleChange):
(WebCore::ScrollAnchoringController::didFindPriorityCandidate):
(WebCore::ScrollAnchoringController::chooseAnchorElement):
(WebCore::ScrollAnchoringController::adjustScrollPositionForAnchoring):
* Source/WebCore/page/scrolling/ScrollAnchoringController.h:
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::styleDidChange):
* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::isInScrollAnchoringAncestorChain const):
(WebCore::RenderObject::setIsInScrollAnchoringAncestorChain):
(WebCore::RenderObject::RenderObjectBitfields::RenderObjectBitfields):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::scrollAnchoringSuppressionStyleDidChange const):
(WebCore::RenderStyle::absolutePositionStyleDidChange const):
* Source/WebCore/rendering/style/RenderStyle.h:

Canonical link: https://commits.webkit.org/270455@main
